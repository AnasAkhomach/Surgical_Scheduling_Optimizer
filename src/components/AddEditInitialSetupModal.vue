<template>
  <div class="modal-overlay" @click.self="handleCancel">
    <div class="modal-content">
      <h3>{{ isEditing ? 'Edit Initial Setup Time' : 'Add New Initial Setup Time' }}</h3>
      <form @submit.prevent="handleSubmit">

        <div class="input-group">
          <label for="setup-type">Surgery Type</label>
          <!-- Dropdown to select surgery type -->
          <!-- This dropdown should be populated by surgery types passed from the parent -->
          <select id="setup-type" v-model="formData.typeId" required :disabled="isEditing"> <!-- Disable type selection when editing -->
            <option value="">Select Surgery Type</option>
            <!-- Options will be populated via v-for on a prop -->
             <option v-for="type in surgeryTypes" :key="type.id" :value="type.id">{{ type.name }}</option>
          </select>
          <small v-if="isEditing">Cannot change Surgery Type when editing.</small>
        </div>

        <div class="input-group">
          <label for="setup-time">Initial Setup Time (minutes)</label>
          <input type="number" id="setup-time" v-model.number="formData.time" required min="0">
        </div>

        <div class="form-actions">
          <button type="submit" class="button-primary">{{ isEditing ? 'Save Changes' : 'Add Setup Time' }}</button>
          <button type="button" class="button-secondary" @click="handleCancel">Cancel</button>
        </div>
      </form>
    </div>
  </div>
</template>

<script setup>
import { ref, watch } from 'vue';

// Define props to receive initial data for editing, editing flag, and the list of surgery types
const props = defineProps({
  initialData: { type: Object, default: null }, // Data for the setup time being edited
  isEditing: { type: Boolean, default: false },
  surgeryTypes: { type: Array, required: true }, // List of surgery types to populate dropdown
});

// Define emits to communicate with parent (e.g., SDSTDataManagementScreen)
const emit = defineEmits(['save', 'cancel']);

// Initialize form data
const formData = ref({
  typeId: null,
  time: 0,
});

// Watch for changes in initialData prop to populate form when editing
watch(() => props.initialData, (newData) => {
  if (newData) {
    // Ensure we don't overwrite typeId if editing
    formData.value = { ...newData };
  } else {
    // Reset form when not editing (e.g., adding new)
     formData.value = { typeId: null, time: 0 };
  }
}, { immediate: true, deep: true }); // Run immediately and watch deeply


const handleSubmit = () => {
  console.log('Submitting Initial Setup Time data:', formData.value);
   // Add validation here if needed (e.g., check if typeId is selected, time is non-negative)
  if (!formData.value.typeId) {
      alert('Please select a Surgery Type.'); // Basic alert, replace with better UI feedback
      return;
  }
   if (formData.value.time < 0) {
       alert('Setup time cannot be negative.'); // Basic alert
       return;
   }

  // Emit the form data to the parent component
  emit('save', formData.value);
  // Reset form (handled by watch effect when parent closes modal/clears initialData)
};

const handleCancel = () => {
  console.log('Cancelling Initial Setup Time form');
  // Emit cancel event to the parent component
  emit('cancel');
  // Reset form (handled by watch effect when parent closes modal/clears initialData)
};

</script>

<style scoped>
.modal-overlay {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5); /* Semi-transparent black background */
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 1000; /* Ensure modal is on top */
}

.modal-content {
  background-color: var(--color-white);
  padding: 30px;
  border-radius: 8px;
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.3);
  max-width: 500px; /* Limit modal width */
  width: 90%;
}

.modal-content h3 {
  margin-top: 0;
  margin-bottom: 20px;
  color: var(--color-very-dark-gray);
  border-bottom: 1px solid var(--color-mid-light-gray);
  padding-bottom: 10px;
}

.input-group {
  margin-bottom: 15px;
}

.input-group label {
  display: block;
  margin-bottom: 5px;
  font-weight: 600;
  color: var(--color-dark-gray);
  font-size: 0.9em;
}

.input-group input[type="number"],
.input-group select {
  width: 100%;
  padding: 8px 12px;
  border: 1px solid var(--color-gray);
  border-radius: 4px;
  box-sizing: border-box; /* Include padding and border in element's total width */
  font-size: 1em;
  color: var(--color-very-dark-gray);
  background-color: var(--color-white);
}

.input-group input:focus,
.input-group select:focus {
    outline: none;
    border-color: var(--color-primary);
    box-shadow: 0 0 0 0.2rem rgba(0, 123, 255, 0.25); /* Example focus style */
}

.input-group small {
     display: block;
     margin-top: 5px;
     color: var(--color-dark-gray);
     font-size: 0.8em;
     font-style: italic;
}

.form-actions {
  margin-top: 20px;
  display: flex;
  justify-content: flex-end; /* Align buttons to the right */
  gap: 10px; /* Space between buttons */
}

.form-actions button {
    padding: 10px 20px;
    font-size: 1em;
}

.button-primary {
     background-color: var(--color-primary);
     color: var(--color-white);
     border: none;
}

.button-primary:hover {
    background-color: var(--color-primary-dark);
}

.button-secondary {
     background-color: var(--color-secondary);
     color: var(--color-white);
     border: none;
}

.button-secondary:hover {
     background-color: #5a6268;
}

</style>