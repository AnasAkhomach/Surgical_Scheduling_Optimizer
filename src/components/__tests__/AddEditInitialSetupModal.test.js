import { describe, it, expect, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import AddEditInitialSetupModal from '../AddEditInitialSetupModal.vue';

describe('AddEditInitialSetupModal', () => {
  const mockSurgeryTypes = [
    { id: 1, name: 'Surgery Type A' },
    { id: 2, name: 'Surgery Type B' },
  ];

  // Test case for adding a new initial setup time
  it('renders correctly in add mode', () => {
    const wrapper = mount(AddEditInitialSetupModal, {
      props: {
        show: true,
        isEditing: false,
        surgeryTypes: mockSurgeryTypes,
      },
    });

    expect(wrapper.find('.modal-overlay').exists()).toBe(true);
    expect(wrapper.text()).toContain('Add New Initial Setup Time');
    expect(wrapper.find('select#setup-type').exists()).toBe(true);
    expect(wrapper.find('select#setup-type').attributes('disabled')).toBeUndefined();
    expect(wrapper.find('input#setup-time').exists()).toBe(true);
    expect(wrapper.find('button[type="submit"]').text()).toBe('Add Setup Time');
    expect(wrapper.find('button[type="button"]').text()).toBe('Cancel');
  });

  // Test case for editing an existing initial setup time
  it('renders correctly in edit mode', () => {
    const initialData = { typeId: 1, time: 30 };
    const wrapper = mount(AddEditInitialSetupModal, {
      props: {
        show: true,
        isEditing: true,
        initialData: initialData,
        surgeryTypes: mockSurgeryTypes,
      },
    });

    expect(wrapper.find('.modal-overlay').exists()).toBe(true);
    expect(wrapper.text()).toContain('Edit Initial Setup Time');
    expect(wrapper.find('select#setup-type').exists()).toBe(true);
    expect(wrapper.find('select#setup-type').attributes('disabled')).toBe(''); // Check for disabled attribute presence
    expect(wrapper.find('select#setup-type').element.value).toBe(initialData.typeId.toString()); // Select value is string
    expect(wrapper.find('input#setup-time').exists()).toBe(true);
    expect(wrapper.find('input#setup-time').element.value).toBe(initialData.time.toString()); // Input value is string
    expect(wrapper.find('button[type="submit"]').text()).toBe('Save Changes');
    expect(wrapper.find('button[type="button"]').text()).toBe('Cancel');
  });

  // Test case for submitting the form in add mode
  it('emits save event with form data when submitting in add mode', async () => {
    const wrapper = mount(AddEditInitialSetupModal, {
      props: {
        show: true,
        isEditing: false,
        surgeryTypes: mockSurgeryTypes,
      },
    });

    const typeSelect = wrapper.find('select#setup-type');
    const timeInput = wrapper.find('input#setup-time');
    const form = wrapper.find('form');

    await typeSelect.setValue(mockSurgeryTypes[0].id);
    await timeInput.setValue(45);
    await form.trigger('submit.prevent');

    expect(wrapper.emitted('save')).toBeTruthy();
    expect(wrapper.emitted('save')[0][0]).toEqual({
      typeId: mockSurgeryTypes[0].id,
      time: 45,
    });
  });

  // Test case for submitting the form in edit mode
  it('emits save event with updated form data when submitting in edit mode', async () => {
    const initialData = { typeId: 1, time: 30, id: 123 }; // Include an ID for editing scenario
    const wrapper = mount(AddEditInitialSetupModal, {
      props: {
        show: true,
        isEditing: true,
        initialData: initialData,
        surgeryTypes: mockSurgeryTypes,
      },
    });

    const timeInput = wrapper.find('input#setup-time');
    const form = wrapper.find('form');

    await timeInput.setValue(60);
    await form.trigger('submit.prevent');

    expect(wrapper.emitted('save')).toBeTruthy();
    expect(wrapper.emitted('save')[0][0]).toEqual({
      typeId: initialData.typeId,
      time: 60,
      id: initialData.id, // Ensure ID is included in edited data
    });
  });

  // Test case for emitting cancel event
  it('emits cancel event when cancel button is clicked', async () => {
    const wrapper = mount(AddEditInitialSetupModal, {
      props: {
         show: true,
         surgeryTypes: mockSurgeryTypes,
      },
    });
    await wrapper.find('button.button-secondary').trigger('click');
    expect(wrapper.emitted('cancel')).toBeTruthy();
  });

  // Test case for clicking outside the modal
   it('emits cancel event when clicking on modal overlay', async () => {
       const wrapper = mount(AddEditInitialSetupModal, {
           props: {
               show: true,
               surgeryTypes: mockSurgeryTypes,
           },
       });
       await wrapper.find('.modal-overlay').trigger('click.self');
       expect(wrapper.emitted('cancel')).toBeTruthy();
   });

  // Test case for not rendering when show is false
  it('does not render when show prop is false', () => {
    const wrapper = mount(AddEditInitialSetupModal, { props: { show: false, surgeryTypes: mockSurgeryTypes } });
    expect(wrapper.find('.modal-overlay').exists()).toBe(false);
  });

  // Snapshot test
  it('matches snapshot in add mode', () => {
    const wrapper = mount(AddEditInitialSetupModal, {
      props: {
        show: true,
        isEditing: false,
        surgeryTypes: mockSurgeryTypes,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

   it('matches snapshot in edit mode', () => {
     const initialData = { typeId: 1, time: 30 };
       const wrapper = mount(AddEditInitialSetupModal, {
         props: {
           show: true,
           isEditing: true,
           initialData: initialData,
           surgeryTypes: mockSurgeryTypes,
         },
       });
       expect(wrapper.html()).toMatchSnapshot();
   });

    // Basic validation tests (can be expanded)
    it('shows alert if surgery type is not selected in add mode', async () => {
        const alertSpy = vi.spyOn(window, 'alert').mockImplementation(() => {}); // Mock alert
        const wrapper = mount(AddEditInitialSetupModal, {
            props: {
                show: true,
                isEditing: false,
                surgeryTypes: mockSurgeryTypes,
            },
        });

        const timeInput = wrapper.find('input#setup-time');
        const form = wrapper.find('form');

        await timeInput.setValue(45);
        await form.trigger('submit.prevent');

        expect(alertSpy).toHaveBeenCalledWith('Please select a Surgery Type.');
        expect(wrapper.emitted('save')).toBeUndefined(); // Ensure save was NOT emitted

        alertSpy.mockRestore(); // Restore original alert
    });

     it('shows alert if time is negative', async () => {
         const alertSpy = vi.spyOn(window, 'alert').mockImplementation(() => {}); // Mock alert
         const wrapper = mount(AddEditInitialSetupModal, {
             props: {
                 show: true,
                 isEditing: false,
                 surgeryTypes: mockSurgeryTypes,
             },
         });

         const typeSelect = wrapper.find('select#setup-type');
         const timeInput = wrapper.find('input#setup-time');
         const form = wrapper.find('form');

         await typeSelect.setValue(mockSurgeryTypes[0].id);
         await timeInput.setValue(-10);
         await form.trigger('submit.prevent');

         expect(alertSpy).toHaveBeenCalledWith('Setup time cannot be negative.');
         expect(wrapper.emitted('save')).toBeUndefined(); // Ensure save was NOT emitted

         alertSpy.mockRestore(); // Restore original alert
     });
});