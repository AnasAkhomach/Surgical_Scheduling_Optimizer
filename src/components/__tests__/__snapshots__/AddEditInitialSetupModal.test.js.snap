// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddEditInitialSetupModal > matches snapshot in add mode 1`] = `
"<div data-v-60ea9dff="" class="modal-overlay">
  <div data-v-60ea9dff="" class="modal-content">
    <h3 data-v-60ea9dff="">Add New Initial Setup Time</h3>
    <form data-v-60ea9dff="">
      <div data-v-60ea9dff="" class="input-group"><label data-v-60ea9dff="" for="setup-type">Surgery Type</label><!-- Dropdown to select surgery type -->
        <!-- This dropdown should be populated by surgery types passed from the parent --><select data-v-60ea9dff="" id="setup-type" required="">
          <!-- Disable type selection when editing -->
          <option data-v-60ea9dff="" value="">Select Surgery Type</option><!-- Options will be populated via v-for on a prop -->
          <option data-v-60ea9dff="" value="1">Surgery Type A</option>
          <option data-v-60ea9dff="" value="2">Surgery Type B</option>
        </select>
        <!--v-if-->
      </div>
      <div data-v-60ea9dff="" class="input-group"><label data-v-60ea9dff="" for="setup-time">Initial Setup Time (minutes)</label><input data-v-60ea9dff="" type="number" id="setup-time" required="" min="0"></div>
      <div data-v-60ea9dff="" class="form-actions"><button data-v-60ea9dff="" type="submit" class="button-primary">Add Setup Time</button><button data-v-60ea9dff="" type="button" class="button-secondary">Cancel</button></div>
    </form>
  </div>
</div>"
`;

exports[`AddEditInitialSetupModal > matches snapshot in edit mode 1`] = `
"<div data-v-60ea9dff="" class="modal-overlay">
  <div data-v-60ea9dff="" class="modal-content">
    <h3 data-v-60ea9dff="">Edit Initial Setup Time</h3>
    <form data-v-60ea9dff="">
      <div data-v-60ea9dff="" class="input-group"><label data-v-60ea9dff="" for="setup-type">Surgery Type</label><!-- Dropdown to select surgery type -->
        <!-- This dropdown should be populated by surgery types passed from the parent --><select data-v-60ea9dff="" id="setup-type" required="" disabled="">
          <!-- Disable type selection when editing -->
          <option data-v-60ea9dff="" value="">Select Surgery Type</option><!-- Options will be populated via v-for on a prop -->
          <option data-v-60ea9dff="" value="1">Surgery Type A</option>
          <option data-v-60ea9dff="" value="2">Surgery Type B</option>
        </select><small data-v-60ea9dff="">Cannot change Surgery Type when editing.</small>
      </div>
      <div data-v-60ea9dff="" class="input-group"><label data-v-60ea9dff="" for="setup-time">Initial Setup Time (minutes)</label><input data-v-60ea9dff="" type="number" id="setup-time" required="" min="0"></div>
      <div data-v-60ea9dff="" class="form-actions"><button data-v-60ea9dff="" type="submit" class="button-primary">Save Changes</button><button data-v-60ea9dff="" type="button" class="button-secondary">Cancel</button></div>
    </form>
  </div>
</div>"
`;
